### DISCUSSION TITLE
Can't get migrated detekt setup to build.gradle.kts to run on build

### DISCUSSION THREAD

**Discussion by mortenholmgaard** [detekt/detekt#3882] [Q&A] [answered]
I have just migrated from Groovy to Kotlin DSL for the build scripts, but I can't get detekt working now. 
I am trying to get detekt to run the the configurations in the detekt {} block on every build, and it works with the previous setup.

Previous setup:
detekt.gradle:
```
apply plugin: "io.gitlab.arturbosch.detekt"

repositories {
    mavenCentral()
    jcenter()
}

dependencies {
    detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:$detekt_version")
}

detekt {
    autoCorrect = true
    toolVersion = "$detekt_version"
    allRules = true
    buildUponDefaultConfig = true
    ignoreFailures = false
    config.setFrom( files("$rootDir/detekt.yml"))
}
```

build.gradle (project)
```
allprojects {
    apply(from = "${rootDir}/detekt.gradle.kts")
    ...
}
```

build.gradle (app)
```
tasks.whenTaskAdded { task ->
    if (task.name == 'compileDebugSources'
        || task.name == 'compileReleaseSources') {
        task.dependsOn detekt
        task.mustRunAfter detekt
    }
}
```

I have tried so many things but latest is this:

build.gradle.kts (projekt)

```
import io.gitlab.arturbosch.detekt.detekt

plugins {
    id("io.gitlab.arturbosch.detekt").version("1.17.1")
}

detekt {
    autoCorrect = true
    toolVersion = "1.17.1"
    allRules = true
    buildUponDefaultConfig = true
    ignoreFailures = false
    config.setFrom(files("$rootDir/detekt.yml"))
}

tasks.withType<io.gitlab.arturbosch.detekt.Detekt>().configureEach {
    jvmTarget = "1.8"
}

...
```

build.gradle.kts (app)

```
tasks.whenTaskAdded {
    if (name == "compileDebugSources"
        || name == "compileReleaseSources"
    ) {
        dependsOn(tasks.withType<io.gitlab.arturbosch.detekt.Detekt>())
        mustRunAfter(tasks.withType<io.gitlab.arturbosch.detekt.Detekt>())
    }
}
```

I am also a bit in doubt if if the detekt {} block is correct as it says it is deprecated with the message from KotlinExtension.kt: "Either apply detekt plugin to root project or follow this advice:
https://docs.gradle.org/current/userguide/kotlin_dsl.html#sec:multi_project_builds_applying_plugins" What to do about that, I honestly don't understand what that means?

**Comment by cortinico** [comment, marked as answer]
Without a reproducer is a bit hard to tell you exactly what's the problem.
I'd suggest you follow the setup from this project:

https://github.com/cortinico/kotlin-android-template/

It uses Gradle Kotlin DSL and setups Detekt on a Gradle Multi-module project.

> What to do about that, I honestly don't understand what that means?

In really short words: 
1. You should
```
plugins {
    id("...") version ... apply false
}
```
on your root `build.gradle.kts` file

2. You should
```
plugins {
    id("...")
}
```
on the `build.gradle.kts` files where you're actually using the plugin (in your case the `app` module). 

I'm also unsure why you're trying to apply to let detekt run before your compilation. Can you clarify there?

**Reply by mortenholmgaard** [reply]
I want to run detekt before compilation just because it is faster to run, so I want those errors right away and not after waiting for a slower compile.
I will take a look at the template there :)

**Reply by mortenholmgaard** [reply]
I have now got it to run detekt(I can see the task in the list when building) with a combination of the above. But it is not producing any errors thouh I have made some errors, that both should tigger error message and autoformat. And when I look at the report I see that it should have run on 500 kt files so I belive that is correct. 
It might not somehow pickup the configuration correctly or something like that?

build.gradle.kts (projekt)

```

plugins {
    id("io.gitlab.arturbosch.detekt") version "1.17.1"
}

tasks.withType<io.gitlab.arturbosch.detekt.Detekt>().configureEach {
    jvmTarget = "1.8"
}

...
```

build.gradle.kts (app)

```

plugins {
    ...
    id("io.gitlab.arturbosch.detekt")
}

detekt {
    autoCorrect = true
    toolVersion = "1.17.1"
    allRules = true
    buildUponDefaultConfig = true
    ignoreFailures = false
    config.setFrom(files("$rootDir/detekt.yml"))
}

tasks.whenTaskAdded {
    if (name == "compileDebugSources"
        || name == "compileReleaseSources"
    ) {
        dependsOn(tasks.detekt)
        mustRunAfter(tasks.detekt)
    }
}
```

**Reply by cortinico** [reply]
> But it is not producing any errors thouh I have made some errors, that both should tigger error message and autoformat

Could you provide a reproducer?

**Reply by mortenholmgaard** [reply]
Here is a reproducer, though as a part of make the reproducer I found that the problem was a bit different than I expected.
Detekt is running fine and also catching some errors, but only very few in compare to before the groovy to kotlin dsl migration.

I the project **MainActivity** there is a method that should trigger a bunch of errors and autoformatting, but it only detects 2 of them. If one enable the detekt plugin in Android Studio it will highlight them, but the detekt tasks as a part of the build will not.
It seems to pickup the **detekt.yml** just fine, but it could seem like the **allRules** not really is triggering correctly somehow.

[DetektExample.zip](https://github.com/detekt/detekt/files/6714109/DetektExample.zip)

One thing I noticed was differnet from the previous setup with groovy is that it dependeded on:
detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:$detekt_version") // notice the "-formatting"

**Reply by cortinico** [reply]
> One thing I noticed was differnet from the previous setup with groovy is that it dependeded on:
detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:$detekt_version") // notice the "-formatting"

That's definitely the cause. A lot of rules related to formatting, as you mention, are inside the `formatting` ruleset (here the list of rules: https://detekt.github.io/detekt/formatting.html).

Such rules are wrapper for ktlint rules and they're enabled only if you add the dependency you mentioned. More on this here: https://detekt.github.io/detekt/extensions.html#integrate-your-extension-with-the-detekt-gradle-plugin

**Reply by mortenholmgaard** [reply]
Thanks - yeah that was it. I wasn't aware that it was different from the other detekt plugin. I just added that line and it all started working!
